### PR TITLE
bpo-8538: Add support for boolean actions to argparse

### PR DESCRIPTION
**Pull request by remilapeyre**
https://bugs.python.org/issue8538

### CODE REVIEW

File: Doc/library/argparse.rst [hunk 1]
```diff
@@ -798,9 +798,19 @@ how the command-line arguments should be handled. The supplied actions are:
...
+is available in ``argparse`` and adds support for boolean actions such as
+``--foo`` and ``--no-foo``::
+
+    >>> import argparse
+    >>> parser = argparse.ArgumentParser()
+    >>> parser.add_argument('--foo', action=argparse.BooleanOptionalAction)
+    >>> parser.parse_args(['--no-foo'])
+    Namespace(foo=False)
+
+The recommended way to do create a custom action is to extend :class:`Action`,
```
Comment (kenahoo) [inline Doc/library/argparse.rst]:
```suggestion
The recommended way to create a custom action is to extend :class:`Action`,
```

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

**Comment by remilapeyre** [comment]
@kenahoo Can you sign the CLA to make @the-knights-who-say-ni happy?

**Comment by kenahoo** [comment]
Thanks @remilapeyre , I've signed it.

**Comment by guludo** [comment]
Hi!

I was about to write a patch to add the same feature. Good thing that I decided to look if someone was already working on it, thank you :-)

I'd like to contribute with the following comments:

## 1. Register action rather than exposing the class

I don't think `BooleanOptionalAction` should be exposed to the module user. Rather, I think it should be "private" (i.e., `_BooleanOptionalAction`) and it should be registered like the other ones (look for calls like `self.register('action', ...)` in the source file). My suggestion is:

```python
self.register('action', 'optional_boolean', _BooleanOptionalAction)
```

## 2. Be aware of short options

Your current implementation does not ignore short options when adding the prefix ``'--no-'``. They should be ignored IMO.

## 3. Ordering or parameters names

I think the ``--no-<option_name>`` variants should be next to ``--<option_name>`` in the  list. Consider the following case where the user decided to have 2 names for the same option:

```python
parser.add_argument('--foo', '--bar', action='optional_boolean', ...)
```

Then, in the help message, ``--foo`` and ``--bar`` would appear before ``--no-foo`` and ``--no-bar``.

## 4. Documentation

In the case of registering the action 'optional_boolean', it would be necessary to update the documentation on the list of supported values for the ``action`` keyword or ``add_argument``.

## 5. My custom action class

About (1) and (2), I have a custom implementation in one project of mine that does that:

```python
class BoolAction(argparse.Action):
    def __init__(self, option_strings, dest, **kw):
        self.original_option_strings = option_strings

        kw['nargs'] = 0

        option_strings = []
        for s in self.original_option_strings:
            option_strings.append(s)

            if s.startswith('--'):
                s = '--no-' + s[2:]
                option_strings.append(s)

        super(BoolAction, self).__init__(option_strings, dest, **kw)

    def __call__(self, parser, namespace, values, option_string):
        value = option_string in self.original_option_strings
        setattr(namespace, self.dest, value)
```

**Comment by kenahoo** [comment]
All of @guludo's comments seem reasonable to me.

**Comment by remilapeyre** [comment]
Hi @guludo, about 1. this has been discussed on b.p.o (https://bugs.python.org/issue8538#msg105620) and Steven Bethard seems to agree (https://bugs.python.org/issue8538#msg166165) so I think it is better to keep it that way before a core reviewers either confirm this API or asks for a registered action. I think it makes 4. unneeded.

Nice catch for 2. and 3. I used something similar than you did for short options. The latest commit fix the ordering too.

**Comment by matrixise** [comment]
I would like to see the result of Travis...

**Review by matrixise — APPROVED** [review]
Thank you for this contribution, I like the feature.

**Comment by matrixise** [comment]
Thank you for your contribution and this feature, really appreciated.

**Comment by remilapeyre** [comment]
Thanks @matrixise !

**Comment by matrixise** [comment]
@remilapeyre with pleasure.

File: Lib/argparse.py [hunk 2]
```diff
@@ -832,9 +833,53 @@ def _get_kwargs(self):
...
+        return self.option_strings[0]
+
     def __call__(self, parser, namespace, values, option_string=None):
         raise NotImplementedError(_('.__call__() not defined'))
 
+class BooleanOptionalAction(Action):
+    def __init__(self,
+                 option_strings,
+                 dest,
+                 const=None,
```
Comment (hauntsaninja) [inline Lib/argparse.py:846]: Is it intentional that we don't pass `const` to super? If so, maybe we shouldn't accept it at all.

Comment (remilapeyre) [reply Lib/argparse.py:846]: This is certainly suspect. I don't remember, I will have a look tomorrow.

Comment (remilapeyre) [reply Lib/argparse.py:846]: Thanks @hauntsaninja, I opened https://bugs.python.org/issue40862 and https://github.com/python/cpython/pull/20623 to fix this issue.

**Pull request merged**